File: algorithms/leetcode/105-construct-binary-tree-from-preorder-and-inorder-traversal/ConstructBinaryTreeFromPreorderAndInorderTraversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    int preorderIndex = 0;
    
    private TreeNode helper(int[] preorder, int[] inorder, int inorderLb, int inorderUb) {
        if (inorderUb - inorderLb <= 0) {
            return null;
        }
        
        TreeNode root = new TreeNode(preorder[preorderIndex++]);
        
        for (int i = inorderLb; i < inorderUb; i++) {
            if (inorder[i] == root.val) {
                root.left = helper(preorder, inorder, inorderLb, i);
                root.right = helper(preorder, inorder, i + 1, inorderUb);
            }
        }
        
        return root;
    }
    
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        return helper(preorder, inorder, 0, inorder.length);
    }
}

File: leetcode/105-construct-binary-tree-from-preorder-and-inorder-traversal/construct-binary-tree-from-preorder-and-inorder-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} preorder
 * @param {number[]} inorder
 * @return {TreeNode}
 */
const buildTree = function (preorder, inorder) {
  const rootNodes = {};
  for (let i = 0; i < inorder.length; i++) {
    rootNodes[inorder[i]] = i;
  }

  const constructTree = (start, end) => {
    const rootVal = preorder[0];

    if (rootNodes[rootVal] >= start && rootNodes[rootVal] <= end) {
      const node = new TreeNode(rootVal);

      preorder.shift();

      node.left = constructTree(start, rootNodes[rootVal] - 1);
      node.right = constructTree(rootNodes[rootVal] + 1, end);

      return node;
    } else {
      return null;
    }
  };

  const head = constructTree(0, inorder.length);

  return head;
};
